src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VsProjectReferenceUtility.cs

```diff
@@ -0,0 +1,437 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using EnvDTE;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.Common;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using VSLangProj;
+using VSLangProj80;
+using EnvDTEProject = EnvDTE.Project;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public static class VSProjectReferenceUtility
+    {
+        /// <summary>
+        /// Returns the closure of all project to project references below this project. While calculating the closure,
+        /// this method populates the <see cref="ExternalProjectReferenceContext.ClosureCache"/> and the
+        /// <see cref="ExternalProjectReferenceContext.DirectReferenceCache"/>.
+        /// </summary>
+        /// <remarks>This uses DTE and should be called from the UI thread.</remarks>
+        public static async Task<IReadOnlyList<ExternalProjectReference>> GetProjectReferenceClosureAsync(
+            EnvDTEProject rootProject,
```

Comment:
`rootDteProject`? I like mentioning root because it implies that this is the starting point for the algorithm executed.